File: src/rankmark/framing.py
"""The v2 self-clocking frame: [ SYNC | HEADER | ECC( PAYLOAD || CRC ) ] repeated.

The decoder knows nothing about where the watermark starts — it slides a
correlator over the carrier bit-stream looking for each profile's SYNC
pattern, reads the tiny repetition-coded header (payload length + namespace
tag), then peels the body back through the ECC stack. The CRC at the center
is the attribution gate: a wrong-lens decode desyncs the entropy gate, and
whatever bits it reads will not checksum-validate.

Profiles own their whole frame geometry (sync length, header repetition,
gate width, ECC stack), because they trade capacity against robustness:

  lean      Barker-13 sync, header x2, CRC16 gate, RS(+2). 71-bit frame for a
            1-byte payload — small enough that any ~150-carrier span of a
            repeated stream contains one whole frame (the Phase-2 milestone).
  standard  31-chip m-sequence, header x3, CRC32 gate, RS(+4) + interleave +
            rate-1/2 convolutional inner code with soft Viterbi.
  robust    header x5, MAGIC16+CRC32 gate (~2^-48 with header consistency),
            RS(+8), deeper interleave.

Deviation from the build plan noted: the plan sketches one fixed geometry
(31-chip sync, header x5). At measured capacity (~63 carrier bits per 500
Qwen tokens at tau=2.0) that geometry cannot meet its own milestone gate, so
geometry moved into the profile and the scan tries each known profile.
"""

import zlib
from dataclasses import dataclass

from .channel import bits_to_bytes, bits_to_int, bytes_to_bits, crc16, int_to_bits
from .ecc import (
    HARD_LLR,
    bits_to_llrs,
    conv_encode,
    CONV_TAIL,
    deinterleave,
    interleave,
    llrs_to_bits,
    rep_decode_soft,
    rep_encode,
    rs_decode,
    rs_encode,
    viterbi_decode,
)

MAGIC = 0xB65D

BARKER_13 = [1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1]
# 31-chip m-sequence from the x^5 + x^2 + 1 LFSR, seed 1.
MSEQ_31 = [1, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1,
           0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0]

LEN_BITS = 6  # payload length in bytes, 1..63
TAG_BITS = 3  # namespace tag: which model family claims this frame
HEADER_BITS = LEN_BITS + TAG_BITS


@dataclass(frozen=True)
class Profile:
    name: str
    sync: tuple[int, ...]
    sync_tol: int  # chips allowed to mismatch during the scan
    header_rep: int
    crc_bytes: int  # 2 = CRC16, 4 = CRC32, 6 = MAGIC16 + CRC32
    rs_nsym: int
    conv: bool
    depth: int  # interleave depth over the RS codeword


PROFILES: dict[int, Profile] = {
    0: Profile("lean", tuple(BARKER_13), 1, 2, 2, 2, False, 1),
    1: Profile("standard", tuple(MSEQ_31), 3, 3, 4, 4, True, 4),
    2: Profile("robust", tuple(MSEQ_31), 5, 5, 6, 8, True, 8),
}
DEFAULT_PROFILE = 1


@dataclass
class DecodedFrame:
    offset: int  # carrier-bit offset where SYNC matched
    payload: bytes
    profile: str
    tag: int
    tag_ok: bool
    sync_errors: int


def tag_of(model_name: str) -> int:
    return crc16(model_name.encode()) & ((1 << TAG_BITS) - 1)


def _checksum(payload: bytes, crc_bytes: int) -> bytes:
    if crc_bytes == 2:
        return crc16(payload).to_bytes(2)
    crc = zlib.crc32(payload).to_bytes(4)
    return crc if crc_bytes == 4 else MAGIC.to_bytes(2) + crc


def _body_coded_bits(payload_len: int, p: Profile) -> int:
    nbits = 8 * (payload_len + p.crc_bytes + p.rs_nsym)
    return 2 * (nbits + CONV_TAIL) if p.conv else nbits


def frame_len_bits(payload_len: int, profile_id: int) -> int:
    p = PROFILES[profile_id]
    return len(p.sync) + HEADER_BITS * p.header_rep + _body_coded_bits(payload_len, p)


def build_frame(payload: bytes, profile_id: int, tag: int) -> list[int]:
    p = PROFILES[profile_id]
    if not 1 <= len(payload) <= (1 << LEN_BITS) - 1:
        raise ValueError(f"payload must be 1..{(1 << LEN_BITS) - 1} bytes")

    body = interleave(rs_encode(payload + _checksum(payload, p.crc_bytes), p.rs_nsym), p.depth)
    body_bits = bytes_to_bits(body)
    if p.conv:
        body_bits = conv_encode(body_bits)

    header = int_to_bits(len(payload), LEN_BITS) + int_to_bits(tag, TAG_BITS)
    return list(p.sync) + rep_encode(header, p.header_rep) + body_bits


def _decode_body(llrs: list[float], payload_len: int, p: Profile) -> bytes | None:
    nbits = 8 * (payload_len + p.crc_bytes + p.rs_nsym)
    bits = viterbi_decode(llrs, nbits) if p.conv else llrs_to_bits(llrs)
    if bits is None:
        return None
    decoded = rs_decode(deinterleave(bits_to_bytes(bits), p.depth), p.rs_nsym)
    if decoded is None:
        return None
    payload = decoded[:payload_len]
    if decoded[payload_len:] != _checksum(payload, p.crc_bytes):
        return None
    return payload


def parse_frames_soft(llrs: list[float], lens_tag: int | None = None) -> list[DecodedFrame]:
    """Scan every offset under every profile geometry; return gate-passing frames.

    Frames whose header tag disagrees with `lens_tag` are kept but flagged
    tag_ok=False — the caller treats them as collision artifacts (a frame
    CRC-passing under lens M but carrying another family's tag is rejected
    from attribution, per the Phase-3 consistency rule).
    """
    hard = llrs_to_bits(llrs)
    frames = []
    for p in PROFILES.values():
        sync_len, hdr_len = len(p.sync), HEADER_BITS * p.header_rep
        for off in range(len(llrs) - sync_len - hdr_len + 1):
            errs = sum(b != s for b, s in zip(hard[off : off + sync_len], p.sync))
            if errs > p.sync_tol:
                continue
            hdr_soft = rep_decode_soft(llrs[off + sync_len : off + sync_len + hdr_len],
                                       p.header_rep)
            hdr = llrs_to_bits(hdr_soft)
            payload_len = bits_to_int(hdr[:LEN_BITS])
            tag = bits_to_int(hdr[LEN_BITS:])
            if payload_len == 0:
                continue
            start = off + sync_len + hdr_len
            end = start + _body_coded_bits(payload_len, p)
            if end > len(llrs):
                continue
            payload = _decode_body(llrs[start:end], payload_len, p)
            if payload is None:
                continue
            frames.append(DecodedFrame(
                offset=off,
                payload=payload,
                profile=p.name,
                tag=tag,
                tag_ok=lens_tag is None or tag == lens_tag,
                sync_errors=errs,
            ))
    return frames


def parse_frames_hard(bits: list[int], lens_tag: int | None = None) -> list[DecodedFrame]:
    return parse_frames_soft(bits_to_llrs(bits), lens_tag)


def frame_layout(payload_len: int, p: Profile) -> list[tuple[str, int]]:
    """(kind, bit-length) parts of one frame, in wire order from its sync.

    Profiles without an inner code expose payload / checksum / parity
    separately; conv or interleaved profiles braid the body, so it is one
    'woven' span.
    """
    parts = [("sync", len(p.sync)), ("header", HEADER_BITS * p.header_rep)]
    if p.conv or p.depth > 1:
        parts.append(("woven", _body_coded_bits(payload_len, p)))
    else:
        parts += [("payload", 8 * payload_len),
                  ("checksum", 8 * p.crc_bytes),
                  ("parity", 8 * p.rs_nsym)]
    return parts


def _spans_from(layout: list[tuple[str, int]], offset: int, limit: int) -> list[dict]:
    spans, cur = [], offset
    for kind, n in layout:
        if cur >= limit:
            break
        spans.append({"kind": kind, "start": cur, "len": min(n, limit - cur)})
        cur += n
    return spans


def frame_spans(frame: DecodedFrame) -> list[dict]:
    """Labelled bit ranges of a fully decoded (checksum-valid) frame."""
    p = next(q for q in PROFILES.values() if q.name == frame.profile)
    layout = frame_layout(len(frame.payload), p)
    total = sum(n for _, n in layout)
    return _spans_from(layout, frame.offset, frame.offset + total)


def partial_spans(llrs: list[float]) -> list[dict]:
    """Tentative labelling of the frame currently arriving at the tail — sync
    the moment it correlates, header once its bits land, and the body regions
    as they stream in, BEFORE the checksum confirms anything.

    Requires an EXACT sync match (not the tolerance parse_frames_soft allows):
    on the author's own lens real syncs are planted exactly, so this lights up
    true frames as they assemble while a random wrong-lens stream shows nothing.
    Display only — parse_frames_soft remains the source of truth.
    """
    hard = llrs_to_bits(llrs)
    n = len(hard)
    best = None  # (offset, profile) — the latest exact sync = frame in progress
    for p in PROFILES.values():
        sync_len = len(p.sync)
        for off in range(n - sync_len, -1, -1):
            if all(b == s for b, s in zip(hard[off : off + sync_len], p.sync)):
                if best is None or off > best[0]:
                    best = (off, p)
                break
    if best is None:
        return []
    off, p = best
    hdr_len = HEADER_BITS * p.header_rep
    hdr_start = off + len(p.sync)
    # can't read the body layout until the whole header has arrived
    if hdr_start + hdr_len > n:
        return _spans_from([("sync", len(p.sync)), ("header", hdr_len)], off, n)
    hdr = llrs_to_bits(rep_decode_soft(llrs[hdr_start : hdr_start + hdr_len], p.header_rep))
    payload_len = bits_to_int(hdr[:LEN_BITS])
    if not 1 <= payload_len <= (1 << LEN_BITS) - 1:
        return _spans_from([("sync", len(p.sync)), ("header", hdr_len)], off, n)
    return _spans_from(frame_layout(payload_len, p), off, n)


def llr_of(rank: int, entropy: float, tau: float) -> float:
    """Per-carrier soft bit from what the lens saw at that position.

    Sign carries the parity (positive = bit 0). Magnitude is a heuristic
    confidence: a marked token sits at rank 0/1, so higher ranks smell like
    edit damage; entropy sitting right on the gate means the carrier/null
    decision itself is fragile.
    """
    conf = HARD_LLR if rank <= 1 else (1.0 if rank < 10 else 0.3)
    conf *= max(0.25, min(1.0, abs(entropy - tau)))
    return conf if rank % 2 == 0 else -conf

File: web/engine/framing.js
// The v2 self-clocking frame: [ SYNC | HEADER | ECC(PAYLOAD || CRC) ] repeated.
// Port of framing.py. build/parse must be bit-exact with the Python side.

import {
  bitsToBytes, bitsToInt, bytesToBits, concatBytes, crc16, crc32, intToBits,
} from "./bits.js";
import {
  bitsToLlrs, convEncode, CONV_TAIL, deinterleave, HARD_LLR, interleave,
  llrsToBits, repDecodeSoft, repEncode, rsDecode, rsEncode, viterbiDecode,
} from "./ecc.js";

const MAGIC = 0xb65d;

const BARKER_13 = [1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1];
const MSEQ_31 = [1, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1,
                 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0];

const BARKER_13_INV = BARKER_13.map(b => 1 - b);   // the copies profile knocks upside down, so a lean frame never answers to it

const LEN_BITS = 6;
const TAG_BITS = 3;
const HEADER_BITS = LEN_BITS + TAG_BITS;

// profile id -> geometry (mirrors PROFILES in framing.py)
export const PROFILES = {
  0: { name: "lean",     sync: BARKER_13, syncTol: 1, headerRep: 2, crcBytes: 2, rsNsym: 2, conv: false, depth: 1 },
  1: { name: "standard", sync: MSEQ_31,   syncTol: 3, headerRep: 3, crcBytes: 4, rsNsym: 4, conv: true,  depth: 4 },
  2: { name: "robust",   sync: MSEQ_31,   syncTol: 5, headerRep: 5, crcBytes: 6, rsNsym: 8, conv: true,  depth: 8 },
  // the lean frame without its repair bytes: short, meant to be repeated, and
  // combined copy by copy at the reader
  3: { name: "copies",   sync: BARKER_13_INV, syncTol: 1, headerRep: 2, crcBytes: 2, rsNsym: 0, conv: false, depth: 1 },
};
export const DEFAULT_PROFILE = 1;

export function tagOf(modelName) {
  return crc16(new TextEncoder().encode(modelName)) & ((1 << TAG_BITS) - 1);
}

function checksum(payload, crcBytes) {
  if (crcBytes === 2) return Uint8Array.of((crc16(payload) >> 8) & 0xff, crc16(payload) & 0xff);
  const c = crc32(payload);
  const crc = Uint8Array.of((c >>> 24) & 0xff, (c >>> 16) & 0xff, (c >>> 8) & 0xff, c & 0xff);
  return crcBytes === 4 ? crc : concatBytes(Uint8Array.of((MAGIC >> 8) & 0xff, MAGIC & 0xff), crc);
}

function bodyCodedBits(payloadLen, p) {
  const nbits = 8 * (payloadLen + p.crcBytes + p.rsNsym);
  return p.conv ? 2 * (nbits + CONV_TAIL) : nbits;
}

export function frameLenBits(payloadLen, profileId) {
  const p = PROFILES[profileId];
  return p.sync.length + HEADER_BITS * p.headerRep + bodyCodedBits(payloadLen, p);
}

export function buildFrame(payload, profileId, tag) {
  const p = PROFILES[profileId];
  if (!(payload.length >= 1 && payload.length <= (1 << LEN_BITS) - 1))
    throw new Error(`payload must be 1..${(1 << LEN_BITS) - 1} bytes`);

  const coded = rsEncode(concatBytes(payload, checksum(payload, p.crcBytes)), p.rsNsym);
  const body = interleave(coded, p.depth);
  let bodyBits = bytesToBits(body);
  if (p.conv) bodyBits = convEncode(bodyBits);

  const header = [...intToBits(payload.length, LEN_BITS), ...intToBits(tag, TAG_BITS)];
  return [...p.sync, ...repEncode(header, p.headerRep), ...bodyBits];
}

function decodeBody(llrs, payloadLen, p) {
  const nbits = 8 * (payloadLen + p.crcBytes + p.rsNsym);
  const bits = p.conv ? viterbiDecode(llrs, nbits) : llrsToBits(llrs);
  if (bits === null) return null;
  const decoded = rsDecode(deinterleave(bitsToBytes(bits), p.depth), p.rsNsym);
  if (decoded === null) return null;
  const payload = decoded.slice(0, payloadLen);
  const gotCrc = decoded.slice(payloadLen);
  const wantCrc = checksum(payload, p.crcBytes);
  if (gotCrc.length !== wantCrc.length) return null;
  for (let i = 0; i < wantCrc.length; i++) if (gotCrc[i] !== wantCrc[i]) return null;
  return payload;
}

// Scan every offset under every profile; return checksum-valid frames. A copy
// that fails on its own is kept as a candidate; candidates of one profile and
// one length are then combined, each bit's confidence summed across copies,
// and the sum decoded (all of them, then each left out in turn, so one false
// knock cannot spoil the rest). A frame reports how many copies it took.
export function parseFramesSoft(llrs, lensTag = null) {
  const hard = llrsToBits(llrs);
  const frames = [];
  for (const p of Object.values(PROFILES)) {
    const syncLen = p.sync.length;
    const hdrLen = HEADER_BITS * p.headerRep;
    const cands = [];
    for (let off = 0; off <= llrs.length - syncLen - hdrLen; off++) {
      let errs = 0;
      for (let i = 0; i < syncLen; i++) if (hard[off + i] !== p.sync[i]) errs++;
      if (errs > p.syncTol) continue;
      const hdrSoft = repDecodeSoft(llrs.slice(off + syncLen, off + syncLen + hdrLen), p.headerRep);
      const hdr = llrsToBits(hdrSoft);
      const payloadLen = bitsToInt(hdr.slice(0, LEN_BITS));
      const tag = bitsToInt(hdr.slice(LEN_BITS));
      if (payloadLen === 0) continue;
      const start = off + syncLen + hdrLen;
      const end = start + bodyCodedBits(payloadLen, p);
      if (end > llrs.length) continue;
      const payload = decodeBody(llrs.slice(start, end), payloadLen, p);
      if (payload === null) { cands.push({ off, start, end, payloadLen, tag, errs }); continue; }
      frames.push({
        offset: off, payload, profile: p.name, tag,
        tagOk: lensTag === null || tag === lensTag, syncErrors: errs, combined: 1,
      });
    }
    const groups = new Map();
    for (const c of cands) { if (!groups.has(c.payloadLen)) groups.set(c.payloadLen, []); groups.get(c.payloadLen).push(c); }
    for (const [payloadLen, g] of groups) {
      if (g.length < 2) continue;
      const bodyLen = g[0].end - g[0].start;
      const decodeSum = subset => {
        const sum = new Array(bodyLen).fill(0);
        for (const c of subset) for (let i = 0; i < bodyLen; i++) sum[i] += llrs[c.start + i];
        return decodeBody(sum, payloadLen, p);
      };
      const subsets = [g];
      if (g.length > 2) for (let i = 0; i < g.length; i++) subsets.push(g.filter((_, j) => j !== i));
      for (const subset of subsets) {
        const payload = decodeSum(subset);
        if (payload === null) continue;
        const a = subset[0];
        frames.push({
          offset: a.off, payload, profile: p.name, tag: a.tag,
          tagOk: lensTag === null || a.tag === lensTag, syncErrors: a.errs, combined: subset.length,
        });
        break;
      }
    }
  }
  return frames;
}

export const parseFramesHard = (bits, lensTag = null) => parseFramesSoft(bitsToLlrs(bits), lensTag);

function frameLayout(payloadLen, p) {
  const parts = [["sync", p.sync.length], ["header", HEADER_BITS * p.headerRep]];
  if (p.conv || p.depth > 1) {
    parts.push(["woven", bodyCodedBits(payloadLen, p)]);
  } else {
    parts.push(["payload", 8 * payloadLen], ["checksum", 8 * p.crcBytes], ["parity", 8 * p.rsNsym]);
  }
  return parts.filter(([, n]) => n > 0);
}

// the segments of a frame about to be written, from bit 0: what the strip draws
export function layoutOf(payloadLen, profileId) {
  const layout = frameLayout(payloadLen, PROFILES[profileId]);
  const total = layout.reduce((s, [, n]) => s + n, 0);
  return spansFrom(layout, 0, total);
}

function spansFrom(layout, offset, limit) {
  const spans = [];
  let cur = offset;
  for (const [kind, n] of layout) {
    if (cur >= limit) break;
    spans.push({ kind, start: cur, len: Math.min(n, limit - cur) });
    cur += n;
  }
  return spans;
}

export function frameSpans(frame) {
  const p = Object.values(PROFILES).find(q => q.name === frame.profile);
  const layout = frameLayout(frame.payload.length, p);
  const total = layout.reduce((s, [, n]) => s + n, 0);
  return spansFrom(layout, frame.offset, frame.offset + total);
}

// Tentative labelling of the frame currently arriving at the tail (exact sync
// only). Display-only — parseFramesSoft stays the source of truth.
export function partialSpans(llrs) {
  const hard = llrsToBits(llrs);
  const n = hard.length;
  let best = null;
  for (const p of Object.values(PROFILES)) {
    const syncLen = p.sync.length;
    for (let off = n - syncLen; off >= 0; off--) {
      let match = true;
      for (let i = 0; i < syncLen; i++) if (hard[off + i] !== p.sync[i]) { match = false; break; }
      if (match) { if (best === null || off > best.off) best = { off, p }; break; }
    }
  }
  if (best === null) return [];
  const { off, p } = best;
  const hdrLen = HEADER_BITS * p.headerRep;
  const hdrStart = off + p.sync.length;
  if (hdrStart + hdrLen > n)
    return spansFrom([["sync", p.sync.length], ["header", hdrLen]], off, n);
  const hdr = llrsToBits(repDecodeSoft(llrs.slice(hdrStart, hdrStart + hdrLen), p.headerRep));
  const payloadLen = bitsToInt(hdr.slice(0, LEN_BITS));
  if (!(payloadLen >= 1 && payloadLen <= (1 << LEN_BITS) - 1))
    return spansFrom([["sync", p.sync.length], ["header", hdrLen]], off, n);
  return spansFrom(frameLayout(payloadLen, p), off, n);
}

export function llrOf(rank, entropy, tau) {
  let conf = rank <= 1 ? HARD_LLR : rank < 10 ? 1.0 : 0.3;
  conf *= Math.max(0.25, Math.min(1.0, Math.abs(entropy - tau)));
  return rank % 2 === 0 ? conf : -conf;
}

export { HEADER_BITS, LEN_BITS, TAG_BITS };
